mesonbuild/interpreter/mesonmain.py

```diff
@@ -277,23 +280,30 @@ def install_dependency_manifest_method(self, args, kwargs):
         self.build.dep_manifest_name = args[0]
 
     @FeatureNew('meson.override_find_program', '0.46.0')
-    @permittedKwargs({})
-    def override_find_program_method(self, args, kwargs):
-        if len(args) != 2:
-            raise InterpreterException('Override needs two arguments')
-        name, exe = args
-        if not isinstance(name, str):
-            raise InterpreterException('First argument must be a string')
-        exe = unholder(exe)
+    @FeatureNewKwargs('meson.override_find_programs', '0.58.0', ['native'])
+    @permittedKwargs({'native'})
+    @typed_pos_args('meson.override_find_programs', str, (ExecutableHolder, mesonlib.File, ExternalProgramHolder))
+    def override_find_program_method(self, args: T.Tuple[str, T.Union[ExecutableHolder, mesonlib.File, ExternalProgramHolder]], kwargs: T.Dict[str, T.Any]):
+        name = args[0]
+        exe: T.Union[build.Executable, mesonlib.File, ExternalProgram] = unholder(args[1])
+
+        # We need native when using a file, and someone might want to override
+        # something for a specific machine, otherwise use the for_machien from
+        # the program itself.
+        if isinstance(exe, mesonlib.File) or 'native' in kwargs:
+            for_machine = self.interpreter.machine_from_native_kwarg(kwargs)
+        else:
+            for_machine = exe.for_machine
```

Comment:
I think we should error out if `native` kwarg does not match `exe.for_machine`.

Reply:
I thought about that too, but sometimes things for the host machine are "runs on the host machine" and sometimes they mean "runs for the host machine", compilers, cmake, pkg-config, and some other tools are all set as "host machine binaries" but they actually run on the build machine and produce output for the host machine. Not sure if that's a worthwile thing to handle or not?

Reply:
Hmmm, the more I think about it the more I think starting with the stricter behavior and relaxing it after we find a usecase might be better.

Reply:
Exactly what I was going to say.